File: trees/binary_tree/full_binary_tree/fb_tree.py
# Full binary tree implementation in Python

# Node class
class Node:
    # Constructor
    def __init__(self, data):
        self.data = data
        self.right = None
        self.left = None


# Check if the binary tree is a full binary tree
def isBinaryTreeFull(root):
    # Check if there's no root, the tree is a full BT
    if (root == None):
        return True
    # Check if the root node or every internal node have no children, we've got a full BT
    if (root.left == None and root.right == None):
        return True
    # Check if the root and every internal node have two children
    if (root.left and root.right):
        return (isBinaryTreeFull(root.left) and isBinaryTreeFull(root.right))
    return False


if __name__ == '__main__':
    # Instantiate the Node class
    root = Node(1)
    root.left = Node(5)
    root.left.left = Node(20)
    root.left.right = Node(25)
    root.right = Node(10)
    root.right.left = Node(90)
    # root.right.right = Node(7)
    if (isBinaryTreeFull(root)):
        print('The binary tree is full')
    else:
        print('The binary tree is not full')
    print()

File: trees/binary_tree/full_binary_tree/fb_tree.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

/* Node struct */
struct Node
{
    int data;
    struct Node *left, *right;
};

/* Function prototyping */

// Create a new node for our tree
struct Node *createNewNode(int data);

// Check if the binary tree is full
bool isBinaryTreeFull(struct Node *root);

int main()
{
    // Let's assume that our tree is empty to get started with
    struct Node *root = NULL;
    root = createNewNode(1);
    root->left = createNewNode(5);
    root->left->left = createNewNode(20);
    root->left->right = createNewNode(25);
    root->right = createNewNode(10);
    root->right->left = createNewNode(30);
    // root->right->right = createNewNode(7);
    if (isBinaryTreeFull(root))
    {
        printf("The binary tree is full\n");
    }
    else
    {
        printf("The binary tree is not full\n");
    }

    /*
                1

            5       10

        20    25  30     7

    */

    return 0;
}

/* Function definition */

struct Node *createNewNode(int data)
{
    // Dynamically allocate memory on the heap for a new node
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    // Assign values
    newNode->data = data;
    newNode->left = newNode->right = NULL;
    return newNode;
}

bool isBinaryTreeFull(struct Node *root)
{
    // Check if there's no root
    if (!root)
    {
        return true;
    }
    // Check if the root has not children
    if (!root->left && !root->right)
    {
        return true;
    }
    // Check if our root has two children and all internal nodes have two children
    if (root->left && root->right)
    {
        return (isBinaryTreeFull(root->left) && isBinaryTreeFull(root->right));
    }
    return false;
}